File: app/src/main/java/de/metzgore/beansplan/dailyschedule/DailyScheduleAdapter.java
package de.metzgore.beansplan.dailyschedule;

import android.databinding.DataBindingUtil;
import android.support.annotation.NonNull;
import android.support.v7.util.DiffUtil;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.ViewGroup;

import java.util.List;
import java.util.Objects;

import de.metzgore.beansplan.R;
import de.metzgore.beansplan.data.room.Reminder;
import de.metzgore.beansplan.data.room.Show;
import de.metzgore.beansplan.data.room.relations.ShowWithReminder;
import de.metzgore.beansplan.databinding.ListItemShowBinding;
import de.metzgore.beansplan.shared.OnReminderButtonClickListener;
import de.metzgore.beansplan.shared.ShowViewHolder;
import de.metzgore.beansplan.shared.ShowViewModel;

public class DailyScheduleAdapter extends RecyclerView.Adapter<ShowViewHolder> {

    private List<ShowWithReminder> shows;
    private OnReminderButtonClickListener listener;

    DailyScheduleAdapter(OnReminderButtonClickListener listener) {
        this.listener = listener;
    }

    public void setShowList(@NonNull final List<ShowWithReminder> showList) {
        if (shows == null) {
            shows = showList;
            notifyItemRangeInserted(0, showList.size());
        } else {
            DiffUtil.DiffResult result = DiffUtil.calculateDiff(new DiffUtil.Callback() {
                @Override
                public int getOldListSize() {
                    return shows.size();
                }

                @Override
                public int getNewListSize() {
                    return showList.size();
                }

                @Override
                public boolean areItemsTheSame(int oldItemPosition, int newItemPosition) {
                    return shows.get(oldItemPosition).show.getId() == showList.get
                            (newItemPosition).show.getId();
                }

                @Override
                public boolean areContentsTheSame(int oldItemPosition, int newItemPosition) {
                    Show newShow = showList.get(newItemPosition).show;
                    Show oldShow = shows.get(oldItemPosition).show;

                    Reminder newReminder = null;
                    Reminder oldReminder = null;

                    if (showList.get(newItemPosition).getReminder() != null) {
                        newReminder = showList.get(newItemPosition).getReminder().get(0);
                    }

                    if (shows.get(oldItemPosition).getReminder() != null) {
                        oldReminder = shows.get(oldItemPosition).getReminder().get(0);
                    }

                    return newShow.equals(oldShow) && Objects.equals(newReminder, oldReminder);
                }
            });
            shows = showList;
            result.dispatchUpdatesTo(this);
        }
    }

    @Override
    @NonNull
    public ShowViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        ListItemShowBinding binding = DataBindingUtil.inflate(LayoutInflater.from(parent
                .getContext()), R.layout.list_item_show, parent, false);
        return new ShowViewHolder(binding);
    }

    @Override
    public void onBindViewHolder(@NonNull ShowViewHolder holder, int position) {
        Reminder reminder = null;

        if (shows.get(position).getReminder() != null) {
            reminder = shows.get(position).getReminder().get(0);
        }

        holder.bind(new ShowViewModel(shows.get(position).show, reminder, listener));
    }

    @Override
    public int getItemCount() {
        return shows == null ? 0 : shows.size();
    }
}

File: app/src/main/java/de/metzgore/beansplan/dailyschedule/DailyScheduleAdapter.kt
package de.metzgore.beansplan.dailyschedule

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.databinding.DataBindingUtil
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView
import de.metzgore.beansplan.R
import de.metzgore.beansplan.data.room.relations.ShowWithReminder
import de.metzgore.beansplan.databinding.ListItemShowBinding
import de.metzgore.beansplan.reminders.RemindersViewModel
import de.metzgore.beansplan.shared.OnReminderButtonClickListener
import de.metzgore.beansplan.shared.ShowViewHolder
import de.metzgore.beansplan.shared.ShowViewModel

class DailyScheduleAdapter(private val remindersViewModel: RemindersViewModel) : RecyclerView.Adapter<ShowViewHolder>() {

    private var showsWithReminder: List<ShowWithReminder>? = null

    fun setShowList(showList: List<ShowWithReminder>) {
        if (showsWithReminder == null) {
            showsWithReminder = showList
            notifyItemRangeInserted(0, showList.size)
        } else {
            //TODO use separate class for here and DailyScheduleAdapter
            val result = DiffUtil.calculateDiff(object : DiffUtil.Callback() {
                override fun getOldListSize(): Int {
                    return showsWithReminder!!.size
                }

                override fun getNewListSize(): Int {
                    return showList.size
                }

                override fun areItemsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
                    return showsWithReminder!![oldItemPosition].show.id == showList[newItemPosition].show.id
                }

                override fun areContentsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
                    val newShow = showList[newItemPosition].show
                    val oldShow = showsWithReminder!![oldItemPosition].show

                    val newReminder = showList[newItemPosition].reminder
                    val oldReminder = showsWithReminder!![oldItemPosition].reminder

                    return newShow == oldShow && newReminder == oldReminder
                }
            })
            showsWithReminder = showList
            result.dispatchUpdatesTo(this)
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ShowViewHolder {
        val binding = DataBindingUtil.inflate<ListItemShowBinding>(LayoutInflater.from(parent
                .context), R.layout.list_item_show, parent, false)
        return ShowViewHolder(binding)
    }

    override fun onBindViewHolder(holder: ShowViewHolder, position: Int) {
        val listener = object : OnReminderButtonClickListener {
            override fun deleteOrUpdateReminder(showWithReminder: ShowWithReminder) {
                remindersViewModel.triggerDeletionOrUpdateDialog(showWithReminder)
            }

            override fun onUpsertReminder(showWithReminder: ShowWithReminder) {
                remindersViewModel.triggerTimePickerDialog(showWithReminder)
            }

            override fun deleteReminder(showWithReminder: ShowWithReminder) {
                // NOOP
            }
        }

        holder.bind(ShowViewModel(showsWithReminder!![position], listener))
    }

    override fun getItemCount(): Int {
        return showsWithReminder?.size ?: 0
    }
}
